rfd/0160-kubernetes-operator-resource-versioning.md

```diff
@@ -0,0 +1,210 @@
+This design adds little changes to the current Teleport Kubernetes Operator
+security model. The only risk is users forgetting to set the `teleportResourceVersion`
+field, in this case we would default to the current version. In a future operator
+version, we can enforce the presence of this field if deemed necessary.
+
+### UX
+
+- This change is backward compatible and offers a straightforward migration
+  path when we introduce a new resource version.
+- The atypical versioning might be confusing for users as it doesn't follow
+  usual Kubernetes patterns.
+
+### API evolution
+
+This design unblocks us today, but does not cover cases in which we introduce a
+new resource version with a completely different structure with fields with
+different types. It is unclear if this can or will happen as this would very
+likely cause other issues in Teleport due to the way we handle resource versioning.
+
+If this were to happen, we would need to revise the solution from this RFD and
+introduce a new resource, or implement conversion hooks. We would also need to
+change the `resources.teleport.dev/vX` version to the current Teleport version.
+
+For example:
+- We implement this RFD in Teleport 15: the API is `resources.teleport.dev/v15`
+- We don't do any breaking changes before v20: the API is still `resources.teleport.dev/v15` for Teleport version 16, 17, 18 and 19.
+- In Teleport v20, we introduce a breaking change in Role. We introduce a new API `resoucres.teleport.dev/v20`.
+
+### Alternatives
+
+When writing this RFD, two alternatives were considered:
+
+#### Conversion hooks
+
+The Kubernetes-friendly approach would be to make the operator aware of how the
+resource is stored in Kubernetes, and do the conversions for every Kubernetes CR
+API call via webhooks.
+
+This alternative has been rejected because of the following reasons:
+- We currently handle resource conversion only from the old version to the newer versions,
+  this would require implementing bidirectional conversion between each resource version
+  (or use a Hub & Spoke approach by converting every version to a common version). This also 
+  raises questions about how to represent non-supported fields during downgrades (e.g. a
+  user gets a Role v7 through a Role v5 API). This would increase the cost of adding new
+  resources and increase the chances of addings bugs in the conversions due to the increasing
+  conversion matrix.
+- We are currently working on hiding `CheckAndSetDefaults()` from the client
+  and consolidating defaults injection and resource conversion server-side
+  This approach is not compatible with the conversion hooks pattern as we'd
+  need to run `CheckAndSetDefaults` in the operator and duplicate the logic.
+- This requires relying on conversion webhooks:
+  - This adds new failure modes: the operator is not healthy, kubernetes cannot talk to the operator
+  - Not all Kubernetes distributions support kube control plane to workload communication by default
```

Comment:
By default GKE Private clusters expose port 443 and 10250. We probably can reuse those ports for the webhook or ensure users add firewall rules to their GKE cluster